port-forwarder/fwd_tcp.go

```diff
@@ -0,0 +1,230 @@
+package port_forwarder
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/slackhq/nebula/service"
+	"golang.org/x/sync/errgroup"
+)
+
+type ForwardConfigOutgoingTcp struct {
+	ForwardConfigOutgoing
+}
+
+func (cfg ForwardConfigOutgoingTcp) ConfigDescriptor() string {
+	return fmt.Sprintf("outbound.tcp.%s.%s", cfg.localListen, cfg.remoteConnect)
+}
+
+type ForwardConfigIncomingTcp struct {
+	ForwardConfigIncoming
+}
+
+func (cfg ForwardConfigIncomingTcp) ConfigDescriptor() string {
+	return fmt.Sprintf("inbound.tcp.%d.%s", cfg.port, cfg.forwardLocalAddress)
+}
+
+type PortForwardingOutgoingTcp struct {
```

Comment:
Suggestion: setup a `context.WithCancel()` here that you then pass down to established sockets (creating a new derived context for all of them), that way live connections will be automatically canceled at shutdown or config changes

Reply:
thanks. done. I also added a test for this specific feature.